### ISSUE TITLE
[Rotarycraft] Autocrafting unit does not work properly with crafting ingedients that don't get used up

### ISSUE THREAD

**Issue by kutschkem** [ReikaKalseki/Reika_Mods_Issues#1489]
I set up a crafting recipe from Immersive Engineering, making iron plates with an engineers hammer. The hammer is not supposed to be used up (just damaged), but is. I tried crafting with the crafting table, it worked as expected, so the issue seems to be the autocrafting unit.

I haven't tried with other recipes, but I suppose the autocrafting unit isn't taking into account that some inputs aren't actually used up in a crafting operation (or however that works code-wise).

**Comment by ReikaKalseki** [comment]
This is quite possibly unfixable, especially if this is implemented with a custom crafting handler.

**Comment by kutschkem** [comment]
I think I understand the problem here is that you have to save the recipe to NBT, and do so by saving the input and output definition. When determining the recipe (in RecipeMode.getRecipe) you already iterate over all the recipes to find the matching one. Can't you do the same when actually crafting, so you have access to the actual crafting handler (for access to the custom getRecipeOutput and getRemainingItems)

**Comment by kutschkem** [comment]
Or, if that's not the same thing, access to the IRecipe.

**Comment by ReikaKalseki** [comment]
Crafting handlers are not accessible or called from IRecipe objects.

**Comment by kutschkem** [comment]
After a bunch of reading, the hammer seems to work like a (fluid) container:

https://github.com/BluSunrize/ImmersiveEngineering/blob/1.7.10/src/main/java/blusunrize/immersiveengineering/common/items/ItemIETool.java

I don't have time right now to check if this is normal forge API or IE API, you might know.

**Comment by kutschkem** [comment]
Never mind, I checked ;-P
The API used is normal Forge API. Item.doesContainerItemLeaveCraftingGrid() and the associated method getContainerItem(ItemStack)

Forget my comments about getRemainingItems, that's post-1.7.10-API.
It appears that this issue should concern all crafting recipes where the container (for example bucket) is not used up.

**Comment by ReikaKalseki** [comment]
That code only has the container item, not the actual damaging code. Meaning that even if I fix it getting consumed, now it will have infinite durability. And *that* sounds like the result of a crafting event or handler.

Also, container items *should* work; buckets *are* left behind.

**Comment by kutschkem** [comment]
... yes it does, or how do you understand this code:

> 	@Override
	public ItemStack getContainerItem(ItemStack stack)
	{
		if(stack.getItemDamage()==0)
		{
			int nbtDamage = ItemNBTHelper.getInt(stack, "hammerDmg")+1;
			if(nbtDamage<hammerMaxDamage)
			{
				ItemStack container = stack.copy();
				ItemNBTHelper.setInt(container, "hammerDmg", nbtDamage);
				return container;
			}
		}
		return null;
	}


I found the code in your craft function where that function is called, but it's hard to follow the control flow through all the different functions - are you calling getContainerItem on the actual ItemStack the recipe is going to use, or a surrogate you store in the recipe? Because for this specific case to work, you would need to call it on the actual ItemStack used.

**Comment by ReikaKalseki** [comment]
> are you calling getContainerItem on the actual ItemStack the recipe is going to use, or a surrogate you store in the recipe?

I will need to check this.

**Comment by ReikaKalseki** [comment]
It does indeed use the cached items' `getContainer`, but there may not be a way to fix this; ME systems, when extracted from, return simply a number of items that were returned, not any specific stack or data therein, nor can I control anything like that. My system for reading available ingredients from a chest behaves similarly, as rarely does the exact item being passed actually matter; two items, all else being equal, are identical.

**Comment by ReikaKalseki** [comment]
I think this was fixed in the last release.
